src/utils.ts

```diff
@@ -303,3 +313,49 @@ export async function createUnsignedFillRelayTransaction(
     { from: simulatedRelayerAddress }
   );
 }
+
+/**
+ * Search back in time for the first block at or older than the desired lookback.
+ *
+ * The approach here is as below:
+ * 1. We use the skip distances, which are meant to be a couple of hours or less for each chain.
+ * 2. Fetch the first block one skip distance from the latest block. Check its timestamp to see how far back it is
+ * relatively to the latest block.
+ * 3. Apply a multiplier to the skip distance based on the time difference and find a block with the desired lookback
+ * This assumes the block production speed is roughly constant.
+ * 4. Check the timestamp again, if we still haven't achieved the desired lookback, repeat step 3 and go back further.
+ * Otherwise, return the block number.
+ *
+ * This is only a rough estimate and can go back further than desired. However, this would minimize the number
+ * of requests and is generally accurate unless the rate of block production has been very volatile over a
+ * short period of time.
+ *
+ * @param provider A valid provider - will be used to find the block with desired lookback.
+ * @param desiredLookback Desired lookback in seconds (e.g. 86400 seconds or 1 day).
+ * @return The block number that's at or older than the desired lookback.
+ */
+export async function findBlockAtOrOlder(
+  provider: providers.Provider | L2Provider<providers.Provider>,
+  desiredLookback: number
+): Promise<number> {
+  const latestBlock = await provider.getBlock("latest");
+  const latestBlockTimestamp = latestBlock.timestamp;
+  const desiredTimestamp = latestBlockTimestamp - desiredLookback;
+  const chainId = (await provider.getNetwork()).chainId;
+  let lastSkipDistance = BlockScanSkipDistances[chainId];
+  let lastBlockNumber = latestBlock.number;
+  let previousLastBlockTimestamp = latestBlockTimestamp;
+  let lastBlockTimestamp = latestBlockTimestamp;
+  while (lastBlockTimestamp > desiredTimestamp) {
+    // Skip the first base case where the last looked up block and the latest block are the same.
+    if (previousLastBlockTimestamp - lastBlockTimestamp > 0) {
+      const lastBlockSpeed = lastSkipDistance / (previousLastBlockTimestamp - lastBlockTimestamp);
```

Comment:
Can `(previousLastBlockTimestamp - lastBlockTimestamp)` ever equal 0?

Reply:
Yes this is the case for the first iteration as they both start out as the latest block's timestamp (no scan has been done).